File: projects/02-industrial-equipment/backend/processor/src/main/java/com/fec/industrial/processor/Handler.java
package com.fec.industrial.processor;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;

import java.net.URI;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Handler implements RequestHandler<SQSEvent, Map<String, Object>> {

    static final String TABLE_NAME = System.getenv().getOrDefault("TABLE_NAME", "fei-readings");
    static DynamoDbClient client;

    // Cached in a static field so warm Lambda invocations reuse the client instead of rebuilding it per SQS batch.
    static synchronized DynamoDbClient client() {
        if (client == null) {
            String endpoint = System.getenv("AWS_ENDPOINT_URL");
            String region = System.getenv().getOrDefault("AWS_REGION", "eu-west-1");
            var builder = DynamoDbClient.builder().region(Region.of(region));
            // Only non-null under LocalStack; a real invocation falls through to its own execution-role credentials.
            if (endpoint != null) {
                builder.endpointOverride(URI.create(endpoint))
                    .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create("test", "test")));
            }
            client = builder.build();
        }
        return client;
    }

    static int persistWindows(List<SQSEvent.SQSMessage> records, DynamoDbClient dynamo, String tableName) {
        int processed = 0;
        for (SQSEvent.SQSMessage record : records) {
            try {
                Map<String, software.amazon.awssdk.services.dynamodb.model.AttributeValue> item =
                    Reshape.toDynamoItem(record.getBody());
                dynamo.putItem(PutItemRequest.builder().tableName(tableName).item(item).build());
                processed++;
            } catch (Exception e) {
                // Fail the whole batch on any bad record so SQS retries it; no partial-batch-failure reporting configured.
                throw new RuntimeException(e);
            }
        }
        return processed;
    }

    @Override
    public Map<String, Object> handleRequest(SQSEvent event, Context context) {
        int processed = persistWindows(event.getRecords(), client(), TABLE_NAME);
        Map<String, Object> result = new HashMap<>();
        result.put("processed", processed);
        return result;
    }
}

File: projects/06-offshore-wind-farm/backend/processor/handler.js
"use strict";

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, PutCommand } = require("@aws-sdk/lib-dynamodb");
const { toItem } = require("./transform");

const TABLE_NAME = process.env.TABLE_NAME || "owf-readings";

let client;
function documentClient() {
  if (client) return client;
  const config = { region: process.env.AWS_REGION || "eu-west-1" };
  // AWS_ENDPOINT_URL is only ever set for LocalStack; a real Lambda always
  // injects AWS_ACCESS_KEY_ID for its own execution-role credentials, so
  // gating on that variable instead would build an incomplete static
  // credential object (missing the session token) and break authentication
  // in a real deployment.
  if (process.env.AWS_ENDPOINT_URL) {
    config.endpoint = process.env.AWS_ENDPOINT_URL;
    config.credentials = {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID || "test",
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || "test",
    };
  }
  client = DynamoDBDocumentClient.from(new DynamoDBClient(config));
  return client;
}

async function writeBatch(records, doc, tableName) {
  let written = 0;
  for (const record of records) {
    await doc.send(new PutCommand({ TableName: tableName, Item: toItem(record.body) }));
    written += 1;
  }
  return written;
}

exports.handler = async (event) => {
  const written = await writeBatch(event.Records || [], documentClient(), TABLE_NAME);
  return { written };
};

exports.writeBatch = writeBatch;
